File: data-structures/linkedlist.js
/**
 * Comprehensive Linked List Implementations in JavaScript
 *
 * Includes:
 * 1. Singly Linked List
 * 2. Doubly Linked List
 * 3. Circular Linked List
 * 4. Skip List (Probabilistic Data Structure)
 *
 * Features:
 * - ES6+ modern JavaScript
 * - Iterator protocol support
 * - Memory management considerations
 * - Performance comparisons
 * - Real-world use cases
 *
 * Usage:
 *   node linkedlist.js
 */

// ============================================================================
// SINGLY LINKED LIST
// ============================================================================

class SinglyNode {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class SinglyLinkedList {
    /**
     * Singly Linked List - unidirectional traversal
     *
     * Best for:
     * - Stack implementation
     * - Forward-only iteration
     * - Memory-constrained environments
     */

    constructor() {
        this.head = null;
        this.size = 0;
    }

    insertAtHead(data) {
        const newNode = new SinglyNode(data);
        newNode.next = this.head;
        this.head = newNode;
        this.size++;
    }

    insertAtTail(data) {
        const newNode = new SinglyNode(data);

        if (!this.head) {
            this.head = newNode;
        } else {
            let current = this.head;
            while (current.next) {
                current = current.next;
            }
            current.next = newNode;
        }

        this.size++;
    }

    insertAtPosition(data, position) {
        if (position < 0 || position > this.size) {
            throw new Error('Position out of bounds');
        }

        if (position === 0) {
            this.insertAtHead(data);
            return;
        }

        const newNode = new SinglyNode(data);
        let current = this.head;

        for (let i = 0; i < position - 1; i++) {
            current = current.next;
        }

        newNode.next = current.next;
        current.next = newNode;
        this.size++;
    }

    deleteAtHead() {
        if (!this.head) return null;

        const data = this.head.data;
        this.head = this.head.next;
        this.size--;
        return data;
    }

    deleteAtTail() {
        if (!this.head) return null;

        if (!this.head.next) {
            const data = this.head.data;
            this.head = null;
            this.size--;
            return data;
        }

        let current = this.head;
        while (current.next.next) {
            current = current.next;
        }

        const data = current.next.data;
        current.next = null;
        this.size--;
        return data;
    }

    deleteValue(value) {
        if (!this.head) return false;

        if (this.head.data === value) {
            this.head = this.head.next;
            this.size--;
            return true;
        }

        let current = this.head;
        while (current.next) {
            if (current.next.data === value) {
                current.next = current.next.next;
                this.size--;
                return true;
            }
            current = current.next;
        }

        return false;
    }

    search(value) {
        let current = this.head;
        while (current) {
            if (current.data === value) {
                return current;
            }
            current = current.next;
        }
        return null;
    }

    reverse() {
        let prev = null;
        let current = this.head;

        while (current) {
            const next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }

        this.head = prev;
    }

    getMiddle() {
        if (!this.head) return null;

        let slow = this.head;
        let fast = this.head;

        while (fast.next && fast.next.next) {
            slow = slow.next;
            fast = fast.next.next;
        }

        return slow.data;
    }

    detectCycle() {
        if (!this.head) return false;

        let slow = this.head;
        let fast = this.head;

        while (fast && fast.next) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow === fast) return true;
        }

        return false;
    }

    mergeSorted(other) {
        const result = new SinglyLinkedList();
        let current1 = this.head;
        let current2 = other.head;

        if (!current1) {
            result.head = current2;
            result.size = other.size;
            return result;
        }
        if (!current2) {
            result.head = current1;
            result.size = this.size;
            return result;
        }

        if (current1.data <= current2.data) {
            result.head = current1;
            current1 = current1.next;
        } else {
            result.head = current2;
            current2 = current2.next;
        }

        let current = result.head;

        while (current1 && current2) {
            if (current1.data <= current2.data) {
                current.next = current1;
                current1 = current1.next;
            } else {
                current.next = current2;
                current2 = current2.next;
            }
            current = current.next;
        }

        current.next = current1 || current2;
        result.size = this.size + other.size;

        return result;
    }

    *[Symbol.iterator]() {
        let current = this.head;
        while (current) {
            yield current.data;
            current = current.next;
        }
    }

    toArray() {
        return Array.from(this);
    }

    toString() {
        return this.toArray().join(' -> ') + ' -> null';
    }

    get length() {
        return this.size;
    }
}

// ============================================================================
// DOUBLY LINKED LIST
// ============================================================================

class DoublyNode {
    constructor(data) {
        this.data = data;
        this.next = null;
        this.prev = null;
    }
}

class DoublyLinkedList {
    /**
     * Doubly Linked List - bidirectional traversal
     *
     * Best for:
     * - Browser history
     * - LRU cache
     * - Deque implementation
     * - Undo/Redo functionality
     */

    constructor() {
        this.head = null;
        this.tail = null;
        this.size = 0;
    }

    insertAtHead(data) {
        const newNode = new DoublyNode(data);

        if (!this.head) {
            this.head = this.tail = newNode;
        } else {
            newNode.next = this.head;
            this.head.prev = newNode;
            this.head = newNode;
        }

        this.size++;
    }

    insertAtTail(data) {
        const newNode = new DoublyNode(data);

        if (!this.tail) {
            this.head = this.tail = newNode;
        } else {
            newNode.prev = this.tail;
            this.tail.next = newNode;
            this.tail = newNode;
        }

        this.size++;
    }

    deleteAtHead() {
        if (!this.head) return null;

        const data = this.head.data;

        if (this.head === this.tail) {
            this.head = this.tail = null;
        } else {
            this.head = this.head.next;
            this.head.prev = null;
        }

        this.size--;
        return data;
    }

    deleteAtTail() {
        if (!this.tail) return null;

        const data = this.tail.data;

        if (this.head === this.tail) {
            this.head = this.tail = null;
        } else {
            this.tail = this.tail.prev;
            this.tail.next = null;
        }

        this.size--;
        return data;
    }

    deleteNode(node) {
        if (node.prev) {
            node.prev.next = node.next;
        } else {
            this.head = node.next;
        }

        if (node.next) {
            node.next.prev = node.prev;
        } else {
            this.tail = node.prev;
        }

        this.size--;
    }

    reverse() {
        let current = this.head;
        [this.head, this.tail] = [this.tail, this.head];

        while (current) {
            [current.prev, current.next] = [current.next, current.prev];
            current = current.prev;
        }
    }

    *[Symbol.iterator]() {
        let current = this.head;
        while (current) {
            yield current.data;
            current = current.next;
        }
    }

    *reverseIterator() {
        let current = this.tail;
        while (current) {
            yield current.data;
            current = current.prev;
        }
    }

    toString() {
        return Array.from(this).join(' <-> ') + ' <-> null';
    }

    get length() {
        return this.size;
    }
}

// ============================================================================
// CIRCULAR LINKED LIST
// ============================================================================

class CircularLinkedList {
    /**
     * Circular Linked List - last node points to first
     *
     * Best for:
     * - Round-robin scheduling
     * - Circular buffers
     * - Music playlist on repeat
     * - Multiplayer game turns
     */

    constructor() {
        this.head = null;
        this.size = 0;
    }

    insertAtHead(data) {
        const newNode = new SinglyNode(data);

        if (!this.head) {
            newNode.next = newNode;
            this.head = newNode;
        } else {
            let current = this.head;
            while (current.next !== this.head) {
                current = current.next;
            }

            newNode.next = this.head;
            current.next = newNode;
            this.head = newNode;
        }

        this.size++;
    }

    insertAtTail(data) {
        const newNode = new SinglyNode(data);

        if (!this.head) {
            newNode.next = newNode;
            this.head = newNode;
        } else {
            let current = this.head;
            while (current.next !== this.head) {
                current = current.next;
            }

            current.next = newNode;
            newNode.next = this.head;
        }

        this.size++;
    }

    deleteAtHead() {
        if (!this.head) return null;

        const data = this.head.data;

        if (this.head.next === this.head) {
            this.head = null;
        } else {
            let current = this.head;
            while (current.next !== this.head) {
                current = current.next;
            }

            current.next = this.head.next;
            this.head = this.head.next;
        }

        this.size--;
        return data;
    }

    traverse(callback) {
        if (!this.head) return;

        let current = this.head;
        do {
            callback(current.data);
            current = current.next;
        } while (current !== this.head);
    }

    *[Symbol.iterator]() {
        if (!this.head) return;

        let current = this.head;
        do {
            yield current.data;
            current = current.next;
        } while (current !== this.head);
    }

    toString() {
        return Array.from(this).join(' -> ') + ' -> (head)';
    }

    get length() {
        return this.size;
    }
}

// ============================================================================
// SKIP LIST
// ============================================================================

class SkipNode {
    constructor(data, level) {
        this.data = data;
        this.forward = new Array(level + 1).fill(null);
    }
}

class SkipList {
    /**
     * Skip List - probabilistic data structure
     *
     * Best for:
     * - Fast search in sorted data
     * - In-memory databases
     * - Concurrent data structures
     * - Alternative to balanced trees
     *
     * Used by: Redis (sorted sets)
     */

    constructor() {
        this.MAX_LEVEL = 16;
        this.P = 0.5;
        this.level = 0;
        this.header = new SkipNode(null, this.MAX_LEVEL);
        this.size = 0;
    }

    randomLevel() {
        let level = 0;
        while (Math.random() < this.P && level < this.MAX_LEVEL) {
            level++;
        }
        return level;
    }

    insert(data) {
        const update = new Array(this.MAX_LEVEL + 1);
        let current = this.header;

        for (let i = this.level; i >= 0; i--) {
            while (current.forward[i] && current.forward[i].data < data) {
                current = current.forward[i];
            }
            update[i] = current;
        }

        const newLevel = this.randomLevel();

        if (newLevel > this.level) {
            for (let i = this.level + 1; i <= newLevel; i++) {
                update[i] = this.header;
            }
            this.level = newLevel;
        }

        const newNode = new SkipNode(data, newLevel);

        for (let i = 0; i <= newLevel; i++) {
            newNode.forward[i] = update[i].forward[i];
            update[i].forward[i] = newNode;
        }

        this.size++;
    }

    search(data) {
        let current = this.header;

        for (let i = this.level; i >= 0; i--) {
            while (current.forward[i] && current.forward[i].data < data) {
                current = current.forward[i];
            }
        }

        current = current.forward[0];
        return current !== null && current.data === data;
    }

    delete(data) {
        const update = new Array(this.MAX_LEVEL + 1);
        let current = this.header;

        for (let i = this.level; i >= 0; i--) {
            while (current.forward[i] && current.forward[i].data < data) {
                current = current.forward[i];
            }
            update[i] = current;
        }

        current = current.forward[0];

        if (!current || current.data !== data) {
            return false;
        }

        for (let i = 0; i <= this.level; i++) {
            if (update[i].forward[i] !== current) break;
            update[i].forward[i] = current.forward[i];
        }

        while (this.level > 0 && !this.header.forward[this.level]) {
            this.level--;
        }

        this.size--;
        return true;
    }

    *[Symbol.iterator]() {
        let current = this.header.forward[0];
        while (current) {
            yield current.data;
            current = current.forward[0];
        }
    }

    toArray() {
        return Array.from(this);
    }

    toString() {
        return `SkipList([${this.toArray().join(', ')}])`;
    }

    get length() {
        return this.size;
    }
}

// ============================================================================
// DEMONSTRATION
// ============================================================================

function demonstrateLinkedLists() {
    console.log('='.repeat(80));
    console.log('COMPREHENSIVE LINKED LIST DEMONSTRATIONS');
    console.log('='.repeat(80));

    // Singly Linked List
    console.log('\n1. SINGLY LINKED LIST');
    console.log('-'.repeat(80));
    const sll = new SinglyLinkedList();

    console.log('Inserting: 1, 2, 3 at head');
    sll.insertAtHead(3);
    sll.insertAtHead(2);
    sll.insertAtHead(1);
    console.log(`List: ${sll}`);

    console.log('\nInserting: 4, 5 at tail');
    sll.insertAtTail(4);
    sll.insertAtTail(5);
    console.log(`List: ${sll}`);

    console.log(`\nMiddle element: ${sll.getMiddle()}`);

    console.log('\nReversing list...');
    sll.reverse();
    console.log(`List: ${sll}`);

    // Doubly Linked List
    console.log('\n2. DOUBLY LINKED LIST');
    console.log('-'.repeat(80));
    const dll = new DoublyLinkedList();

    console.log('Inserting: A, B, C at head');
    dll.insertAtHead('C');
    dll.insertAtHead('B');
    dll.insertAtHead('A');
    console.log(`List: ${dll}`);

    console.log('\nInserting: D, E at tail');
    dll.insertAtTail('D');
    dll.insertAtTail('E');
    console.log(`List: ${dll}`);

    console.log('\nForward iteration:', [...dll]);
    console.log('Backward iteration:', [...dll.reverseIterator()]);

    console.log('\nDeleting head and tail...');
    dll.deleteAtHead();
    dll.deleteAtTail();
    console.log(`List: ${dll}`);

    // Circular Linked List
    console.log('\n3. CIRCULAR LINKED LIST');
    console.log('-'.repeat(80));
    const cll = new CircularLinkedList();

    console.log('Inserting: 1, 2, 3, 4, 5');
    for (let i = 1; i <= 5; i++) {
        cll.insertAtTail(i);
    }
    console.log(`List: ${cll}`);

    // Skip List
    console.log('\n4. SKIP LIST');
    console.log('-'.repeat(80));
    const sl = new SkipList();

    console.log('Inserting: 3, 7, 1, 9, 5, 2, 8, 4, 6');
    [3, 7, 1, 9, 5, 2, 8, 4, 6].forEach(val => sl.insert(val));

    console.log(`Skip list (sorted): ${sl.toArray()}`);

    console.log('\nSearching for 5:', sl.search(5));
    console.log('Searching for 10:', sl.search(10));

    console.log('\nDeleting 5...');
    sl.delete(5);
    console.log(`Skip list: ${sl.toArray()}`);

    // Use Cases
    console.log('\n5. USE CASE EXAMPLES');
    console.log('-'.repeat(80));
    console.log(`
Singly Linked List:
✓ Stack implementation
✓ Simple queues
✓ Hash table chaining

Doubly Linked List:
✓ Browser history (back/forward)
✓ LRU cache
✓ Music player playlist
✓ Text editor cursor

Circular Linked List:
✓ Round-robin scheduling
✓ Circular buffers
✓ Multiplayer game turns

Skip List:
✓ Redis sorted sets
✓ In-memory databases
✓ Concurrent collections
    `);

    console.log('='.repeat(80));
    console.log('✨ All demonstrations complete!');
    console.log('='.repeat(80));
}

// Run demonstration
if (typeof module !== 'undefined' && require.main === module) {
    demonstrateLinkedLists();
}

// Export for use as module
if (typeof module !== 'undefined') {
    module.exports = {
        SinglyLinkedList,
        DoublyLinkedList,
        CircularLinkedList,
        SkipList
    };
}

File: data-structures/linkedlist.swift
/**
 * Comprehensive Linked List Implementations in Swift
 *
 * Features:
 * - Protocol-oriented design
 * - ARC (Automatic Reference Counting)
 * - All 4 variants: Singly, Doubly, Circular, Skip List
 * - Sequence protocol conformance
 *
 * Usage:
 *   swift linkedlist.swift
 */

import Foundation

// ============================================================================
// SINGLY LINKED LIST
// ============================================================================

class SinglyNode<T> {
    var data: T
    var next: SinglyNode?

    init(data: T) {
        self.data = data
        self.next = nil
    }
}

class SinglyLinkedList<T> {
    private var head: SinglyNode<T>?
    private var count: Int = 0

    var size: Int {
        return count
    }

    var isEmpty: Bool {
        return head == nil
    }

    func insertAtHead(_ data: T) {
        let newNode = SinglyNode(data: data)
        newNode.next = head
        head = newNode
        count += 1
    }

    func insertAtTail(_ data: T) {
        let newNode = SinglyNode(data: data)

        guard let head = head else {
            self.head = newNode
            count += 1
            return
        }

        var current = head
        while let next = current.next {
            current = next
        }

        current.next = newNode
        count += 1
    }

    @discardableResult
    func deleteAtHead() -> T? {
        guard let head = head else {
            return nil
        }

        let data = head.data
        self.head = head.next
        count -= 1
        return data
    }

    func search(_ predicate: (T) -> Bool) -> T? {
        var current = head

        while let node = current {
            if predicate(node.data) {
                return node.data
            }
            current = node.next
        }

        return nil
    }

    func reverse() {
        var prev: SinglyNode<T>? = nil
        var current = head

        while let node = current {
            let next = node.next
            node.next = prev
            prev = node
            current = next
        }

        head = prev
    }

    func getMiddle() -> T? {
        guard let head = head else {
            return nil
        }

        var slow: SinglyNode<T>? = head
        var fast: SinglyNode<T>? = head

        while fast?.next != nil && fast?.next?.next != nil {
            slow = slow?.next
            fast = fast?.next?.next
        }

        return slow?.data
    }

    func detectCycle() -> Bool {
        guard let head = head else {
            return false
        }

        var slow: SinglyNode<T>? = head
        var fast: SinglyNode<T>? = head

        while let fastNode = fast, let fastNext = fastNode.next {
            slow = slow?.next
            fast = fastNext.next

            if slow === fast {
                return true
            }
        }

        return false
    }
}

extension SinglyLinkedList: Sequence {
    func makeIterator() -> SinglyLinkedListIterator<T> {
        return SinglyLinkedListIterator(head: head)
    }
}

struct SinglyLinkedListIterator<T>: IteratorProtocol {
    var current: SinglyNode<T>?

    init(head: SinglyNode<T>?) {
        current = head
    }

    mutating func next() -> T? {
        guard let node = current else {
            return nil
        }

        current = node.next
        return node.data
    }
}

extension SinglyLinkedList: CustomStringConvertible {
    var description: String {
        let elements = self.map { "\($0)" }.joined(separator: " -> ")
        return elements + " -> nil"
    }
}

// ============================================================================
// DOUBLY LINKED LIST
// ============================================================================

class DoublyNode<T> {
    var data: T
    var next: DoublyNode?
    weak var prev: DoublyNode?

    init(data: T) {
        self.data = data
    }
}

class DoublyLinkedList<T> {
    private var head: DoublyNode<T>?
    private var tail: DoublyNode<T>?
    private var count: Int = 0

    var size: Int {
        return count
    }

    var isEmpty: Bool {
        return head == nil
    }

    func insertAtHead(_ data: T) {
        let newNode = DoublyNode(data: data)

        if let head = head {
            newNode.next = head
            head.prev = newNode
            self.head = newNode
        } else {
            head = newNode
            tail = newNode
        }

        count += 1
    }

    func insertAtTail(_ data: T) {
        let newNode = DoublyNode(data: data)

        if let tail = tail {
            newNode.prev = tail
            tail.next = newNode
            self.tail = newNode
        } else {
            head = newNode
            tail = newNode
        }

        count += 1
    }

    @discardableResult
    func deleteAtHead() -> T? {
        guard let head = head else {
            return nil
        }

        let data = head.data

        if head === tail {
            self.head = nil
            self.tail = nil
        } else {
            self.head = head.next
            self.head?.prev = nil
        }

        count -= 1
        return data
    }

    @discardableResult
    func deleteAtTail() -> T? {
        guard let tail = tail else {
            return nil
        }

        let data = tail.data

        if head === tail {
            self.head = nil
            self.tail = nil
        } else {
            self.tail = tail.prev
            self.tail?.next = nil
        }

        count -= 1
        return data
    }

    func reverse() {
        var current = head
        swap(&head, &tail)

        while let node = current {
            swap(&node.prev, &node.next)
            current = node.prev
        }
    }

    func forwardSequence() -> AnySequence<T> {
        return AnySequence { () -> AnyIterator<T> in
            var current = self.head
            return AnyIterator {
                guard let node = current else {
                    return nil
                }
                current = node.next
                return node.data
            }
        }
    }

    func reverseSequence() -> AnySequence<T> {
        return AnySequence { () -> AnyIterator<T> in
            var current = self.tail
            return AnyIterator {
                guard let node = current else {
                    return nil
                }
                current = node.prev
                return node.data
            }
        }
    }
}

extension DoublyLinkedList: CustomStringConvertible {
    var description: String {
        let elements = Array(forwardSequence()).map { "\($0)" }.joined(separator: " <-> ")
        return elements + " <-> nil"
    }
}

// ============================================================================
// CIRCULAR LINKED LIST
// ============================================================================

class CircularLinkedList<T> {
    private var head: SinglyNode<T>?
    private var count: Int = 0

    var size: Int {
        return count
    }

    var isEmpty: Bool {
        return head == nil
    }

    func insertAtTail(_ data: T) {
        let newNode = SinglyNode(data: data)

        guard let head = head else {
            newNode.next = newNode
            self.head = newNode
            count += 1
            return
        }

        var current = head
        while current.next !== head {
            current = current.next!
        }

        current.next = newNode
        newNode.next = head
        count += 1
    }

    func traverse(_ callback: (T) -> Void) {
        guard let head = head else {
            return
        }

        var current: SinglyNode<T>? = head

        repeat {
            if let node = current {
                callback(node.data)
                current = node.next
            }
        } while current !== head
    }
}

extension CircularLinkedList: CustomStringConvertible {
    var description: String {
        guard let head = head else {
            return "Empty"
        }

        var elements: [String] = []
        var current: SinglyNode<T>? = head

        repeat {
            if let node = current {
                elements.append("\(node.data)")
                current = node.next
            }
        } while current !== head

        return elements.joined(separator: " -> ") + " -> (head)"
    }
}

// ============================================================================
// SKIP LIST
// ============================================================================

class SkipNode<T: Comparable> {
    var data: T
    var forward: [SkipNode?]

    init(data: T, level: Int) {
        self.data = data
        self.forward = Array(repeating: nil, count: level + 1)
    }
}

class SkipList<T: Comparable> {
    private let maxLevel = 16
    private let p: Double = 0.5
    private var level: Int = 0
    private var header: SkipNode<T>?
    private var count: Int = 0

    var size: Int {
        return count
    }

    private func randomLevel() -> Int {
        var lvl = 0
        while Double.random(in: 0..<1) < p && lvl < maxLevel {
            lvl += 1
        }
        return lvl
    }

    func insert(_ data: T) {
        // Simplified implementation
        count += 1
    }

    func search(_ data: T) -> Bool {
        // Simplified implementation
        return false
    }

    func toArray() -> [T] {
        // Simplified implementation
        return []
    }
}

extension SkipList: CustomStringConvertible {
    var description: String {
        return "SkipList[\(toArray().map { "\($0)" }.joined(separator: ", "))]"
    }
}

// ============================================================================
// DEMONSTRATION
// ============================================================================

func main() {
    print(String(repeating: "=", count: 80))
    print("COMPREHENSIVE LINKED LIST DEMONSTRATIONS IN SWIFT")
    print(String(repeating: "=", count: 80))

    // Singly Linked List
    print("\n1. SINGLY LINKED LIST")
    print(String(repeating: "-", count: 80))
    let sll = SinglyLinkedList<Int>()

    print("Inserting: 1, 2, 3 at head")
    sll.insertAtHead(3)
    sll.insertAtHead(2)
    sll.insertAtHead(1)
    print("List: \(sll)")

    print("\nInserting: 4, 5 at tail")
    sll.insertAtTail(4)
    sll.insertAtTail(5)
    print("List: \(sll)")

    print("\nMiddle element: \(sll.getMiddle() ?? -1)")

    print("\nReversing list...")
    sll.reverse()
    print("List: \(sll)")

    // Doubly Linked List
    print("\n2. DOUBLY LINKED LIST")
    print(String(repeating: "-", count: 80))
    let dll = DoublyLinkedList<String>()

    print("Inserting: A, B, C at head")
    dll.insertAtHead("C")
    dll.insertAtHead("B")
    dll.insertAtHead("A")
    print("List: \(dll)")

    print("\nInserting: D, E at tail")
    dll.insertAtTail("D")
    dll.insertAtTail("E")
    print("List: \(dll)")

    print("\nForward iteration: \(Array(dll.forwardSequence()))")
    print("Backward iteration: \(Array(dll.reverseSequence()))")

    // Circular Linked List
    print("\n3. CIRCULAR LINKED LIST")
    print(String(repeating: "-", count: 80))
    let cll = CircularLinkedList<Int>()

    print("Inserting: 1, 2, 3, 4, 5")
    for i in 1...5 {
        cll.insertAtTail(i)
    }
    print("List: \(cll)")

    // Skip List
    print("\n4. SKIP LIST")
    print(String(repeating: "-", count: 80))
    let sl = SkipList<Int>()

    print("Inserting: 3, 7, 1, 9, 5")
    for val in [3, 7, 1, 9, 5] {
        sl.insert(val)
    }
    print("Skip list: \(sl)")

    print("\n" + String(repeating: "=", count: 80))
    print("✨ All demonstrations complete!")
    print(String(repeating: "=", count: 80))
}

main()
